oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/StorageID.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    public StorageID appendMutant(final String[] source, final String value) {
+        if (sealed) {
+            throw new IllegalStateException("The storage ID is sealed. Can't append a new fragment, source=" + source);
+        }
+        fragments.add(new Fragment(source, String.class, true, value));
+        return this;
+    }
+
+    /**
+     * @return the string ID concatenating the values of {@link #fragments} by the underline(_).
+     */
+    public String build() {
+        sealed = true;
+        if (builtID == null) {
+            builtID = Joiner.on(Const.ID_CONNECTOR).join(fragments);
+        }
+        return builtID;
+    }
+
+    /**
+     * @return a read-only list to avoid unexpected change for metric ID.
+     */
+    public List<Fragment> read() {
+        sealed = true;
+        return Collections.unmodifiableList(fragments);
+    }
+
+    @RequiredArgsConstructor
+    @Getter
+    public static class Fragment {
+        /**
+         * The column name of the value, or the original column names of the mutate value.
+         *
+         * The names could be
+         * 1. Always one column if this is not {@link #mutate} and from a certain persistent column.
+         * 2. Be null if {@link #mutate} is true and no relative column, such as the original value is not in
+         * the persistence.
+         * 3. One or multi-values if the value is built through a symmetrical or asymmetrical encoding algorithm.
+         */
+        private final String[] name;
+        /**
+         * Represent the class type of the {@link #value}.
+         */
+        private final Class<?> type;
+        /**
+         * If true, the field was from {@link #name}, and value is changed by internal rules.
+         * Such as time bucket downsampling, use a day-level time-bucket to build the ID for a minute dimension metric.
+         */
+        private final boolean mutate;
+        private final Object value;
+
+        public Optional<String[]> getName() {
+            return Optional.ofNullable(name);
+        }
+
+        @Override
+        public String toString() {
+            return value.toString();
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            final Fragment fragment = (Fragment) o;
+
+            // Probably incorrect - comparing Object[] arrays with Arrays.equals
+            if (!Arrays.equals(name, fragment.name))
+                return false;
+            if (value != null ? !value.equals(fragment.value) : fragment.value != null)
+                return false;
+
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = Arrays.hashCode(name);
+            result = 31 * result + (value != null ? value.hashCode() : 0);
+            return result;
+        }
```

Comment:
The reason why you have to write these instead of using Lombok `EqualsAndHashCode(of = {"name", "value"})` is that, you have a `getName` that returns an `Optional<String[]>`, Lombok can compare array correctly but not when arrays are wrapped into `Optional`, but luckily you can change the behaviour of Lombok and not to use the getter, but using this :


```java
    @EqualsAndHashCode(of = {
        "name",
        "value"
    }, doNotUseGetters = true)
```